### ISSUE TITLE
crashed inside openssl method

### ISSUE THREAD

**Issue by g1204** [emscripten-core/emscripten#15115]
in rand_pool_acquire_entropy it try to call 'open' and crashed.
I tried **-s FORCE_FILESYSTEM=1**, yet not working.

![图片](https://user-images.githubusercontent.com/11961269/134613643-632b43c4-7b4f-49f1-966f-c546b18ef881.png)

**Comment by sbc100** [comment]
Can you try building the latest version of emscripten, and also with `-sASSERTIONS` enabled (of `-O0` will also enable assertions)?

The `resolveSymbol` function should print a message when it fails.

It looks like you are using at least one SIDE_MODULE along with a MAIN_MODULE?  Is that right?  Can you share the link commands you are using for both of these?   Have you tried doing a static build with MAIN_MODULE or SIDE_MODULE?  It can be useful for debugging purposed at least to have static build even if you plan to deploy dynamic libraries.

**Comment by g1204** [comment]
After adding _open to EXPORTED_FUNCTIONS, it fixed.
However, now it crash at 'clock_gettime', and cannot be fixed either by adding it to EXPORTED_FUNCTIONS or DEFAULT_LIBRARY_FUNCS_TO_INCLUDE. (These problems doesn't exist for static linking, that is, no MAIN_MODULE or SIDE_MODULE, build all sources together).

Uncaught (in promise) RuntimeError: abort(Assertion failed: undefined symbol `clock_gettime`. perhaps a side module was not linked in? if this global was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment) at Error
    at jsStackTrace (gsdk.js:9)
    at stackTrace (gsdk.js:9)
    at abort (gsdk.js:9)
    at assert (gsdk.js:9)
    at resolveSymbol (gsdk.js:9)
    at obj.<computed> (gsdk.js:9)
    at <anonymous>:wasm-function[4284]:0x17aef0

**Comment by g1204** [comment]
EXPORT_ALL=1 fixed the `clock_gettime` error.
Still try to find a setting to export clock_gettime.

**Comment by g1204** [comment]
Solved.

By adding `clock_gettime` to DEFAULT_LIBRARY_FUNCS_TO_INCLUDE and also add `_clock_gettime` to EXPORTED_FUNCTIONS.

**Comment by sbc100** [comment]
I think you must be somehow using dynamic linking because otherwise all the needed functions would be included at link time automatically.  

In fact, the `resolveSymbols` function is only available if `MAIN_MODULE` is defined at link time:

https://github.com/emscripten-core/emscripten/blob/12b2b675c202d2abc0a3f6c756939dde39eb199e/src/library_dylink.js#L458

Its inside this giant `#if MAIN_MODULE` block:
https://github.com/emscripten-core/emscripten/blob/12b2b675c202d2abc0a3f6c756939dde39eb199e/src/library_dylink.js#L177

**Comment by g1204** [comment]
Yes. The problem was about dynamic linking(MAIN_MODULE & SIDE_MODULE).
Seems some system methods which the SIDE_MODULE depend on must be specified in EXPORTED_FUNCTIONS and DEFAULT_LIBRARY_FUNCS_TO_INCLUDE when compiling MAIN_MODULE.

**Comment by sbc100** [comment]
There are two different modes for MAIN_MODULE.   With MAIN_MODULE=1 everything is always included so you shouldn't need to export anything explicitly.. but you will end up with a huge module because it will include all functions.   If you use MAIN_MODULE=2 the system will only include what it thinks is needed and DCE the rest... in this mode you can either specify the dependency manually using `EXPORTED_FUNCTIONS` and DEFAULT_LIBRARY_FUNCS_TO_INCLUDE` or you can pass the side module on the command line and it will include precisely what is needed and nothing more automatically.

In short, the best method is to use MAIN_MODULE=1 + specify the side module on the command line.     In this mode the side module will normally be loaded automatically on startup.  If you don't want that you can add `-sNO_AUTOLOAD_DYLIBS`

**Comment by g1204** [comment]
Awesome suggestion. Thanks a lot!
